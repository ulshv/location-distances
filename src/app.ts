import getDistance from 'geolib/es/getDistance';

import { ILocation, ILocationPairWithDistance } from "./typings";
import { parseRequestData, createPairsFromList } from "./utils";
import { getLocationDetailsMock } from "./api";

export const calculateResponseData = async (locationsQueriesRaw: string[]) => {
  const locationQueries = parseRequestData(locationsQueriesRaw);
  const locations: ILocation[] = [];

  for (let locationQuery of locationQueries) {
    const location = await getLocationDetailsMock(locationQuery);
    if (location) locations.push(location);
  }

  const locationPairs = createPairsFromList(locations);

  const locationPairsWithDistances: ILocationPairWithDistance[] = locationPairs.map(pair => ({
    pair,
    distance: getDistance(
      { latitude: pair[0].latitude, longitude: pair[0].longitude },
      { latitude: pair[1].latitude, longitude: pair[1].longitude }
    )
  }));

  const pairsRelatedToLocationByQuery: { [key: string]: ILocationPairWithDistance[] } = {};

  locations.forEach(location => {
    pairsRelatedToLocationByQuery[location.query] = [];
  });

  locationPairsWithDistances.forEach(pairWithDistance => {
    pairWithDistance.pair.forEach(location => {
      pairsRelatedToLocationByQuery[location.query].push(pairWithDistance);
    });
  });

  const closestLocationPairsWithDistances = locations.map(location =>
    pairsRelatedToLocationByQuery[location.query].sort((a, b) => a.distance - b.distance)[0]
  );

  return closestLocationPairsWithDistances.map((pairWithDistance, index) => {
    const from = pairWithDistance.pair.find(location => location === locations[index]);
    const to   = pairWithDistance.pair.find(location => location !== locations[index]);

    return ({
      from,
      to,
      distance: (pairWithDistance.distance / 1000).toFixed(2) + ' km',
    })
  })
}
