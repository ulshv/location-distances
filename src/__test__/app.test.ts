import { testFn } from "../app";

describe('app', () => {
  it('should return true', () => {
    expect(testFn()).toBe(true);
  })
})
