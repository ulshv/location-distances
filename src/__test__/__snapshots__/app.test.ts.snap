// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app.ts Should correctly get location pairs and calculate distances 1`] = `
Array [
  Object {
    "distance": "635.70 km",
    "from": Object {
      "latitude": 55.75583,
      "longitude": 37.61778,
      "name": "Moscow",
      "query": "moscow",
    },
    "to": Object {
      "latitude": 59.95,
      "longitude": 30.31667,
      "name": "Saint Petersburg",
      "query": "saint petersburg",
    },
  },
  Object {
    "distance": "128.73 km",
    "from": Object {
      "latitude": 40.7648,
      "longitude": -73.9808,
      "name": "New York",
      "query": "new york",
    },
    "to": Object {
      "latitude": 40.0115,
      "longitude": -75.1327,
      "name": "Philadelphia",
      "query": "philadelphia",
    },
  },
  Object {
    "distance": "128.73 km",
    "from": Object {
      "latitude": 40.0115,
      "longitude": -75.1327,
      "name": "Philadelphia",
      "query": "philadelphia",
    },
    "to": Object {
      "latitude": 40.7648,
      "longitude": -73.9808,
      "name": "New York",
      "query": "new york",
    },
  },
  Object {
    "distance": "635.70 km",
    "from": Object {
      "latitude": 59.95,
      "longitude": 30.31667,
      "name": "Saint Petersburg",
      "query": "saint petersburg",
    },
    "to": Object {
      "latitude": 55.75583,
      "longitude": 37.61778,
      "name": "Moscow",
      "query": "moscow",
    },
  },
]
`;
